// @agentlend/contracts/typechain-types/contracts/CollateralManager.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BigNumberish,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../common";

export interface CollateralManagerInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "LIQUIDATION_PENALTY_BPS"
      | "MAX_COLLATERAL_TOKENS"
      | "activeTokenCount"
      | "calculateRequiredCollateral"
      | "collateralBalances"
      | "collateralRatios"
      | "collateralTokens"
      | "depositCollateral"
      | "getCollateralBalance"
      | "getCollateralRatio"
      | "getTotalCollateral"
      | "lendingPool"
      | "owner"
      | "releaseCollateral"
      | "renounceOwnership"
      | "scoreThresholds"
      | "seizeCollateral"
      | "setCollateralToken"
      | "setLendingPool"
      | "totalCollateral"
      | "transferOwnership"
      | "withdrawCollateral"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "CollateralDeposited"
      | "CollateralSeized"
      | "CollateralTokenUpdated"
      | "CollateralWithdrawn"
      | "OwnershipTransferred"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "LIQUIDATION_PENALTY_BPS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "MAX_COLLATERAL_TOKENS",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "activeTokenCount",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "calculateRequiredCollateral",
    values: [BigNumberish, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "collateralBalances",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "collateralRatios",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "collateralTokens",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "depositCollateral",
    values: [BytesLike, AddressLike, BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getCollateralBalance",
    values: [BytesLike, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "getCollateralRatio",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "getTotalCollateral",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "lendingPool",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "releaseCollateral",
    values: [BytesLike, AddressLike, BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "renounceOwnership",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "scoreThresholds",
    values: [BigNumberish]
  ): string;
  encodeFunctionData(
    functionFragment: "seizeCollateral",
    values: [BytesLike, AddressLike, BigNumberish, AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "setCollateralToken",
    values: [AddressLike, boolean]
  ): string;
  encodeFunctionData(
    functionFragment: "setLendingPool",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "totalCollateral",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "withdrawCollateral",
    values: [BytesLike, AddressLike, BigNumberish]
  ): string;

  decodeFunctionResult(
    functionFragment: "LIQUIDATION_PENALTY_BPS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "MAX_COLLATERAL_TOKENS",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "activeTokenCount",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "calculateRequiredCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "collateralBalances",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "collateralRatios",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "collateralTokens",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "depositCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getCollateralBalance",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getCollateralRatio",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "getTotalCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "lendingPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "releaseCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "renounceOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "scoreThresholds",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "seizeCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setCollateralToken",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "setLendingPool",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "totalCollateral",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "withdrawCollateral",
    data: BytesLike
  ): Result;
}

export namespace CollateralDepositedEvent {
  export type InputTuple = [
    agentDID: BytesLike,
    token: AddressLike,
    amount: BigNumberish
  ];
  export type OutputTuple = [agentDID: string, token: string, amount: bigint];
  export interface OutputObject {
    agentDID: string;
    token: string;
    amount: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CollateralSeizedEvent {
  export type InputTuple = [
    agentDID: BytesLike,
    token: AddressLike,
    amount: BigNumberish,
    liquidator: AddressLike
  ];
  export type OutputTuple = [
    agentDID: string,
    token: string,
    amount: bigint,
    liquidator: string
  ];
  export interface OutputObject {
    agentDID: string;
    token: string;
    amount: bigint;
    liquidator: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CollateralTokenUpdatedEvent {
  export type InputTuple = [token: AddressLike, active: boolean];
  export type OutputTuple = [token: string, active: boolean];
  export interface OutputObject {
    token: string;
    active: boolean;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace CollateralWithdrawnEvent {
  export type InputTuple = [
    agentDID: BytesLike,
    token: AddressLike,
    amount: BigNumberish
  ];
  export type OutputTuple = [agentDID: string, token: string, amount: bigint];
  export interface OutputObject {
    agentDID: string;
    token: string;
    amount: bigint;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OwnershipTransferredEvent {
  export type InputTuple = [previousOwner: AddressLike, newOwner: AddressLike];
  export type OutputTuple = [previousOwner: string, newOwner: string];
  export interface OutputObject {
    previousOwner: string;
    newOwner: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface CollateralManager extends BaseContract {
  connect(runner?: ContractRunner | null): CollateralManager;
  waitForDeployment(): Promise<this>;

  interface: CollateralManagerInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  LIQUIDATION_PENALTY_BPS: TypedContractMethod<[], [bigint], "view">;

  MAX_COLLATERAL_TOKENS: TypedContractMethod<[], [bigint], "view">;

  activeTokenCount: TypedContractMethod<[], [bigint], "view">;

  calculateRequiredCollateral: TypedContractMethod<
    [loanAmount: BigNumberish, score: BigNumberish],
    [bigint],
    "view"
  >;

  collateralBalances: TypedContractMethod<
    [arg0: BytesLike, arg1: AddressLike],
    [bigint],
    "view"
  >;

  collateralRatios: TypedContractMethod<[arg0: BigNumberish], [bigint], "view">;

  collateralTokens: TypedContractMethod<
    [arg0: AddressLike],
    [
      [string, boolean, bigint] & {
        token: string;
        active: boolean;
        decimals: bigint;
      }
    ],
    "view"
  >;

  depositCollateral: TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  getCollateralBalance: TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike],
    [bigint],
    "view"
  >;

  getCollateralRatio: TypedContractMethod<
    [score: BigNumberish],
    [bigint],
    "view"
  >;

  getTotalCollateral: TypedContractMethod<
    [token: AddressLike],
    [bigint],
    "view"
  >;

  lendingPool: TypedContractMethod<[], [string], "view">;

  owner: TypedContractMethod<[], [string], "view">;

  releaseCollateral: TypedContractMethod<
    [
      agentDID: BytesLike,
      token: AddressLike,
      amount: BigNumberish,
      recipient: AddressLike
    ],
    [void],
    "nonpayable"
  >;

  renounceOwnership: TypedContractMethod<[], [void], "nonpayable">;

  scoreThresholds: TypedContractMethod<[arg0: BigNumberish], [bigint], "view">;

  seizeCollateral: TypedContractMethod<
    [
      agentDID: BytesLike,
      token: AddressLike,
      amount: BigNumberish,
      liquidator: AddressLike
    ],
    [void],
    "nonpayable"
  >;

  setCollateralToken: TypedContractMethod<
    [token: AddressLike, active: boolean],
    [void],
    "nonpayable"
  >;

  setLendingPool: TypedContractMethod<
    [_lendingPool: AddressLike],
    [void],
    "nonpayable"
  >;

  totalCollateral: TypedContractMethod<[arg0: AddressLike], [bigint], "view">;

  transferOwnership: TypedContractMethod<
    [newOwner: AddressLike],
    [void],
    "nonpayable"
  >;

  withdrawCollateral: TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "LIQUIDATION_PENALTY_BPS"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "MAX_COLLATERAL_TOKENS"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "activeTokenCount"
  ): TypedContractMethod<[], [bigint], "view">;
  getFunction(
    nameOrSignature: "calculateRequiredCollateral"
  ): TypedContractMethod<
    [loanAmount: BigNumberish, score: BigNumberish],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "collateralBalances"
  ): TypedContractMethod<
    [arg0: BytesLike, arg1: AddressLike],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "collateralRatios"
  ): TypedContractMethod<[arg0: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "collateralTokens"
  ): TypedContractMethod<
    [arg0: AddressLike],
    [
      [string, boolean, bigint] & {
        token: string;
        active: boolean;
        decimals: bigint;
      }
    ],
    "view"
  >;
  getFunction(
    nameOrSignature: "depositCollateral"
  ): TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "getCollateralBalance"
  ): TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike],
    [bigint],
    "view"
  >;
  getFunction(
    nameOrSignature: "getCollateralRatio"
  ): TypedContractMethod<[score: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "getTotalCollateral"
  ): TypedContractMethod<[token: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "lendingPool"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "owner"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "releaseCollateral"
  ): TypedContractMethod<
    [
      agentDID: BytesLike,
      token: AddressLike,
      amount: BigNumberish,
      recipient: AddressLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "renounceOwnership"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "scoreThresholds"
  ): TypedContractMethod<[arg0: BigNumberish], [bigint], "view">;
  getFunction(
    nameOrSignature: "seizeCollateral"
  ): TypedContractMethod<
    [
      agentDID: BytesLike,
      token: AddressLike,
      amount: BigNumberish,
      liquidator: AddressLike
    ],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setCollateralToken"
  ): TypedContractMethod<
    [token: AddressLike, active: boolean],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "setLendingPool"
  ): TypedContractMethod<[_lendingPool: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "totalCollateral"
  ): TypedContractMethod<[arg0: AddressLike], [bigint], "view">;
  getFunction(
    nameOrSignature: "transferOwnership"
  ): TypedContractMethod<[newOwner: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "withdrawCollateral"
  ): TypedContractMethod<
    [agentDID: BytesLike, token: AddressLike, amount: BigNumberish],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "CollateralDeposited"
  ): TypedContractEvent<
    CollateralDepositedEvent.InputTuple,
    CollateralDepositedEvent.OutputTuple,
    CollateralDepositedEvent.OutputObject
  >;
  getEvent(
    key: "CollateralSeized"
  ): TypedContractEvent<
    CollateralSeizedEvent.InputTuple,
    CollateralSeizedEvent.OutputTuple,
    CollateralSeizedEvent.OutputObject
  >;
  getEvent(
    key: "CollateralTokenUpdated"
  ): TypedContractEvent<
    CollateralTokenUpdatedEvent.InputTuple,
    CollateralTokenUpdatedEvent.OutputTuple,
    CollateralTokenUpdatedEvent.OutputObject
  >;
  getEvent(
    key: "CollateralWithdrawn"
  ): TypedContractEvent<
    CollateralWithdrawnEvent.InputTuple,
    CollateralWithdrawnEvent.OutputTuple,
    CollateralWithdrawnEvent.OutputObject
  >;
  getEvent(
    key: "OwnershipTransferred"
  ): TypedContractEvent<
    OwnershipTransferredEvent.InputTuple,
    OwnershipTransferredEvent.OutputTuple,
    OwnershipTransferredEvent.OutputObject
  >;

  filters: {
    "CollateralDeposited(bytes32,address,uint256)": TypedContractEvent<
      CollateralDepositedEvent.InputTuple,
      CollateralDepositedEvent.OutputTuple,
      CollateralDepositedEvent.OutputObject
    >;
    CollateralDeposited: TypedContractEvent<
      CollateralDepositedEvent.InputTuple,
      CollateralDepositedEvent.OutputTuple,
      CollateralDepositedEvent.OutputObject
    >;

    "CollateralSeized(bytes32,address,uint256,address)": TypedContractEvent<
      CollateralSeizedEvent.InputTuple,
      CollateralSeizedEvent.OutputTuple,
      CollateralSeizedEvent.OutputObject
    >;
    CollateralSeized: TypedContractEvent<
      CollateralSeizedEvent.InputTuple,
      CollateralSeizedEvent.OutputTuple,
      CollateralSeizedEvent.OutputObject
    >;

    "CollateralTokenUpdated(address,bool)": TypedContractEvent<
      CollateralTokenUpdatedEvent.InputTuple,
      CollateralTokenUpdatedEvent.OutputTuple,
      CollateralTokenUpdatedEvent.OutputObject
    >;
    CollateralTokenUpdated: TypedContractEvent<
      CollateralTokenUpdatedEvent.InputTuple,
      CollateralTokenUpdatedEvent.OutputTuple,
      CollateralTokenUpdatedEvent.OutputObject
    >;

    "CollateralWithdrawn(bytes32,address,uint256)": TypedContractEvent<
      CollateralWithdrawnEvent.InputTuple,
      CollateralWithdrawnEvent.OutputTuple,
      CollateralWithdrawnEvent.OutputObject
    >;
    CollateralWithdrawn: TypedContractEvent<
      CollateralWithdrawnEvent.InputTuple,
      CollateralWithdrawnEvent.OutputTuple,
      CollateralWithdrawnEvent.OutputObject
    >;

    "OwnershipTransferred(address,address)": TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;
    OwnershipTransferred: TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;
  };
}
